src/main/java/com/amazon/ion/impl/bin/IonRawBinaryWriter_1_1.kt

```diff
@@ -114,50 +140,155 @@ class IonRawBinaryWriter_1_1 internal constructor(
+                buffer.writeByte(OpCodes.ANNOTATIONS_1_FLEX_SYM)
+                annotationsTotalLength += buffer.writeFlexSym(utf8StringEncoder.encode(annotationsTextBuffer[0].toString()))
+                annotationsTextBuffer[0] = null
+            }
+            -3 -> {
+                buffer.writeByte(OpCodes.ANNOTATIONS_2_FLEX_SYM)
+                annotationsTotalLength += writeFlexSymFromAnnotationsBuffer(0)
+                annotationsTotalLength += writeFlexSymFromAnnotationsBuffer(1)
+            }
+            else -> annotationsTotalLength += writeManyAnnotations()
         }
+        currentContainer.length += annotationsTotalLength
+
         numAnnotations = 0
+        annotationFlexSymFlag = 0
         hasFieldName = false
         valueWriterExpression()
     }
 
+    /**
+     * Writes a FlexSym annotation for the specified position in the annotations buffers.
+     */
+    private fun writeFlexSymFromAnnotationsBuffer(i: Int): Int {
+        val annotationText = annotationsTextBuffer[i]
+        return if (annotationText != null) {
+            annotationsTextBuffer[i] = null
+            buffer.writeFlexSym(utf8StringEncoder.encode(annotationText.toString()))
+        } else {
+            buffer.writeFlexSym(annotationsIdBuffer[i])
+        }
+    }
+
+    /**
+     * Writes 3 or more annotations for SIDs or FlexSyms
+     */
+    private fun writeManyAnnotations(): Long {
+        currentContainer = containerStack.push {
+            it.clear()
+            it.type = Annotations
+            it.position = buffer.position()
+        }
+        if (annotationFlexSymFlag == FLEX_SYMS_REQUIRED) {
+            buffer.writeByte(OpCodes.ANNOTATIONS_MANY_FLEX_SYM)
+            buffer.reserve(lengthPrefixPreallocation)
+            for (i in 0 until numAnnotations) {
+                currentContainer.length += writeFlexSymFromAnnotationsBuffer(i)
+            }
+        } else {
+            buffer.writeByte(OpCodes.ANNOTATIONS_MANY_SYMBOL_ADDRESS)
+            buffer.reserve(lengthPrefixPreallocation)
+            for (i in 0 until numAnnotations) {
+                currentContainer.length += buffer.writeFlexUInt(annotationsIdBuffer[i])
+            }
+        }
+
+        val numAnnotationsBytes = currentContainer.length
+        val numLengthPrefixBytes = writeCurrentContainerLength()
+
+        // Set the new current container
+        containerStack.pop()
+        currentContainer = containerStack.peek()
+
+        return numLengthPrefixBytes + numAnnotationsBytes
+    }
+
     /**
      * Helper function for finishing scalar values. Similar concerns for containers are handled in [stepOut].
      */
     private inline fun closeScalar(valueWriterExpression: () -> Int) {
         val numBytesWritten = valueWriterExpression()
-
-        // Update the container length (unless it's Top)
-        if (currentContainer.type != Top) currentContainer.length += numBytesWritten
+        currentContainer.length += numBytesWritten
```

Comment:
🗺️  A long is pretty big, and it doesn't hurt anyone to have the length of `Top` being tracked, so I got rid of the branching here (and similarly in `stepOut()`).